### PR TITLE
[WIP] Fixed #20461 -- Added support for running Django's test suite in parallel

### PR DESCRIPTION
**Pull request by aaugustin**

### CODE REVIEW

**Comment by aaugustin** [comment]
Trac ticket: #20461

**Comment by collinanderson** [comment]
a quick FYI on py 3.4: ImportError: No module named 'tblib'

**Comment by charettes** [comment]
@collinanderson, it's a [thid party application](https://pypi.python.org/pypi/tblib/0.2.0).

**Comment by aaugustin** [comment]
Indeed you need to `pip install tblib`. That's the easiest solution I could find to get tracebacks from the multiprocessing pool.

**Comment by collinanderson** [comment]
Not bad :)

Ran 9003 tests in 61.995s

FAILED (failures=1, errors=31, skipped=429, expected failures=6)
Destroying test database for alias 'default'...
Destroying test database for alias 'other'...
Command exited with non-zero status 1
289.78user 13.87system 1:32.20elapsed 329%CPU (0avgtext+0avgdata 360600maxresident)k
8296inputs+336464outputs (72major+3018357minor)pagefaults 0swaps

**Comment by aaugustin** [comment]
That's a lot of errors... I don't get that many.

**Comment by collinanderson** [comment]
https://gist.github.com/collinanderson/55fdb3081f43a1f7636f

**Comment by akaariai** [comment]
We likely need to mark some tests as single-core only. For example, we don't want to run multiple memcached servers on different ports so that each core has its own memcached instance... Similarly for selenium tests, and tests that play with filesystem.

**Comment by aaugustin** [comment]
@akaariai I'd rather make these tests independent from one another. We can use different filesystem paths, different key prefixes, etc.

**Comment by akaariai** [comment]
That is of course a better solution.

**Comment by aaugustin** [comment]
Well, it turns out that the better solution isn't always viable. I have implemented a `SerializeMixin` that makes it easy to enforce serial execution of test cases inheriting a common mixin.

**Comment by aaugustin** [comment]
The test failure is a fluke: the checkout from GitHub failed.

File: django/test/runner.py [hunk 1]
```diff
@@ -52,19 +54,109 @@ def printErrorList(self, flavour, errors):
...
+            except StopIteration:
+                pool.close()
+                break
+
+            errors.append(test_errors)
+            stats.append(test_stats)
+
+        pool.join()
+
+        # Gross hack to trick result into displaying the right data.
```
Comment (berkerpeksag) [inline django/test/runner.py]: You can move these lines into a TextTestResult subclass then use it in RemoteTextTestRunner by overriding the resultclass attribute (the call to self._makeResult() in RemoteTextTestRunner.run() should handle the rest). The multiprocessing code can be moved into RemoteTextTestRunner.run() in my opinion.

Comment (aaugustin) [inline django/test/runner.py]: > You can move these lines into a TextTestResult subclass then use it in RemoteTextTestRunner by overriding the resultclass attribute.

Indeed. I tried not to add too many classes but your proposal would be more clean.

> The multiprocessing code can be moved into RemoteTextTestRunner.run() in my opinion.

I'm not sure I understand. `RemoteTextTestRunner` executes on the workers, not on the master.

**Comment by aaugustin** [comment]
Closing in favor of some new PRs.

**Pull request closed without merging**